Artifact fails to download with `invalid certificate: BadDER`

My company uses SSL inspection as part of its security envelope. This system requires the equivalent of a self-signed intermediate certificate to be added to the system certificate store as a root certificate authority in order for any SSL based connection to function. Any SSL based connection will be intercepted and rewritten; the new connection will be signed by this CA instead of its original authority and packets will be rewritten to match.

I've been unable to use pants in this context, which, when attempting to download it's artifacts, reports the following error:

```
Exception: Error downloading file: error sending request for url (https://github.com/pantsbuild/pex/releases/download/v2.1.35/pex): error trying to connect: invalid certificate: BadDER
```

With the help of some very nice folks on slack, we've already traced this issue to the rust part of the system, in the `rustls` crate used by `reqwest`. I've attatched the relevant logs, and I'd be happy to generate / capture anything else you'd find helpful. Long story short, it looks like I'm running into the issue reported here: https://github.com/ctz/rustls/issues/127. Rustls diverges from openssl behaviour; when attempting to process the certificate, it attempts to read the `subjectAltName` from it. In my case, the final dynamic inspection certificate has one, but the two previous signing CA certificates in chain only have a common name. Since this extension is missing, when `rustls` attempts to read the field, it fails, implodes, and the request fails with the above error (or so we currently think).

The certificates probably should have subjectAltName extensions set. However, they're generated dynamically, and not really in my control - they're handled at a much higher level by a different department in a large organization; it will take months if not years to get something like that fixed (if I even can get someone to care). Most people don't run into this issue because openssl is more permissive in this case and doesn't fault when the subjectAltName is missing. So if it's possible, it would be really helpful if a pants side fix could be implemented to address this.

SSL Intercept systems like this are becoming increasingly common at large corporations, to prevent in-network malicious actors from messaging home in SSL packets. I promise you this isn't a one-off issue.

I'm a full stack developer, and normally I'd just figure out the solution, and issue an MR with a proposed fix (since this affects me, it seems like the right thing to do). However, I'm not at all familiar with rust, and this is a security issue - I'm not the person you want poking around in the source code trying to fix this. I'd be happy to help in any other way I can, and I can test / debug / log anything that would be helpful.

I have captured the SSL exchange at the packet level, and I have confirmed that the certificates in the exchange are fully formed and valid. As the certificates are authorities, and could be used for signing purposes, I can't actually share any of these, but, I can report back any information that would be helpful, or try to capture any details that might be useful.

Thanks in advance for any help :smile:

[20210501194000_pants_debug.log](https://github.com/pantsbuild/pants/files/6410708/20210501194000_pants_debug.log)

Well...I spoke too soon. It's only half working. My company (for reasons I won't get into) hard cuts our internet mainline every night so there are limits on the testing I can do against external targets.

Removing the culprit certificates does let me set `PANTS_CA_CERTS_PATH` to my custom ca bundle without pantsd just immediately crashing, so we've fixed that (yay!). However, I'm still getting the same error:

[20210503094100_pants_debug.log](https://github.com/pantsbuild/pants/files/6415167/20210503094100_pants_debug.log)

This seems to have brought things to parity wether or not I set `PANTS_CA_CERTS_PATH` at this point 🤔. It still seems to crash down in `rustls` in a way I can't quite understand and doesn't seem to get captured by the logs (I've grabbed a snippet):

```
09:40:55.45 [DEBUG] (reqwest::connect) starting new connection: https://github.com/
09:40:55.45 [DEBUG] (hyper::client::connect::dns) resolving host="github.com" 
09:40:55.45 [DEBUG] (process_execution) Running Searching for `bash` on PATH=/usr/bin:/bin:/usr/local/bin under semaphore with concurrency id: 1
09:40:55.45 [DEBUG] (process_execution::local) Obtaining exclusive spawn lock for process with argv ["./find_binary.sh", "bash"] since we materialized its executable RelativePath("find_binary.sh").
09:40:55.46 [DEBUG] (process_execution::local) spawned local process as Some(8415) for Process { argv: ["./find_binary.sh", "bash"], env: {"PATH": "/usr/bin:/bin:/usr/local/bin"}, working_directory: None, input_files: Digest { hash: Fingerprint<124495cfdd1160d88a797da1b5de65f0b9d57642b553d7ff5195d5375411b492>, size_bytes: 91 }, output_files: {}, output_directories: {}, timeout: None, execution_slot_variable: None, description: "Searching for `bash` on PATH=/usr/bin:/bin:/usr/local/bin", level: Debug, append_only_caches: {}, jdk_home: None, platform_constraint: None, is_nailgunnable: false, cache_scope: PerRestart }
09:40:55.47 [DEBUG] (hyper::client::connect::http) connecting to 140.82.114.4:443 
09:40:55.50 [DEBUG] (hyper::client::connect::http) connected to 140.82.114.4:443 
09:40:55.50 [DEBUG] (rustls::client::hs) No cached session for DNSNameRef("github.com")
09:40:55.50 [DEBUG] (rustls::client::hs) Not resuming any session
09:40:55.54 [DEBUG] (rustls::client::hs) ALPN protocol is None
09:40:55.54 [DEBUG] (rustls::client::hs) Using ciphersuite TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384
09:40:55.54 [DEBUG] (rustls::client::tls12) ECDHE curve is ECParameters { curve_type: NamedCurve, named_group: secp384r1 }
09:40:55.54 [DEBUG] (rustls::client::tls12) Server DNS name is DNSName("github.com")
09:40:55.54 [WARN] (rustls::session) Sending fatal alert DecodeError
09:40:55.54 [DEBUG] (workunit_store) Completed: Downloading: DownloadFile(url='https://github.com/pantsbuild/pex/releases/download/v2.1.35/pex', expected_digest=FileDigest(fingerprint='3b1c97ebd79a8b650c81f0afbdbefef17c3d1a128219a8e95b5f35c2c9dfd53... (37 characters truncated)
09:40:55.54 [DEBUG] (workunit_store) Completed: pants.core.util_rules.external_tool.download_external_tool
09:40:55.54 [DEBUG] (workunit_store) Canceled: Find Python interpreter to bootstrap PEX
09:40:55.54 [DEBUG] (workunit_store) Completed: Find Python interpreter for constraints
09:40:55.54 [DEBUG] (workunit_store) Canceled: Finding the `bash` binary
09:40:55.55 [DEBUG] (workunit_store) Completed: `dependencies` goal
09:40:55.55 [DEBUG] (pants.engine.internals.scheduler) computed 1 nodes in 0.156098 seconds. there are 250 total nodes.
09:40:55.55 [ERROR] (pants.base.exception_sink) Exception caught: (pants.engine.internals.scheduler.ExecutionError)
```

Regardless of logging level, there doesn't seem to ever be any additional information regarding the warning `09:40:55.54 [WARN] (rustls::session) Sending fatal alert DecodeError`, but that seems to be the crux of the issue. Thoughts? I'm still setup for testing if there is more you'd like me to try...

--- a/src/rust/Cargo.toml
+++ b/src/rust/Cargo.toml
@@ -199,7 +199,7 @@ pyo3 = { version = "0.28.3", features = ["parking_lot"] }
 pyo3-build-config = "0.28.3"
 rand = "0.10.0"
 regex = "1.12.3"
-reqwest = { version = "0.12", default-features = false }
+reqwest = { version = "0.13", default-features = false }
 rlimit = "0.10"
 rustls = "0.23.37"
 rustls-native-certs = "0.8.3"

--- a/src/rust/engine/Cargo.toml
+++ b/src/rust/engine/Cargo.toml
@@ -54,7 +54,7 @@ process_execution = { path = "../process_execution" }
 pyo3 = { workspace = true }
 rand = { workspace = true }
 regex = { workspace = true }
-reqwest = { workspace = true, default-features = false, features = ["stream", "rustls-tls"] }
+reqwest = { workspace = true, default-features = false, features = ["stream", "rustls"] }
 rule_graph = { path = "../rule_graph" }
 sandboxer = { path = "../process_execution/sandboxer" }
 smallvec = { workspace = true, features = ["union"] }

